### PR TITLE
better error handling for get_build_config_by_labels_filtered

### PR DESCRIPTION
**Pull request by mlangsdorf**
raise exceptions if no filter match is found and there's only
one response, or if there's no filter key and only one response.

Signed-off-by: Mark Langsdorf <mlangsdo@redhat.com>

### CODE REVIEW

File: osbs/core.py [hunk 1]
```diff
@@ -373,6 +373,15 @@ def get_build_config_by_labels_filtered(self, label_selectors, filter_key, filte
             raise OsbsException(
                 "More than one build config found for labels: %r" %
                 (label_selectors, ))
+        elif filter_value:
+            raise OsbsException(
+                "No match for {0} found for labels: %r".format(filter_value) %
+                (label_selectors, ))
```
Comment (twaugh) [inline osbs/core.py]: Should we also raise an exception if there is more than one match after applying the filter? That's what the docstring says we do.

I feel like the function should be structured a bit more like:
- get the build configs that match the label selectors
- apply the filter, if there is a filter to apply
- raise an exception if we don't have a single build config
- return it

Comment (mlangsdorf) [reply osbs/core.py]: Restructured as requested.

**Comment by mlangsdorf** [comment]
Fixed.

File: osbs/core.py [hunk 2]
```diff
@@ -360,19 +360,24 @@ def get_build_config_by_labels_filtered(self, label_selectors, filter_key, filte
...
-                "Build config not found for labels: %r" %
-                (label_selectors, ))
         if filter_value is not None:
+            build_configs = []
             for build_config in items:
                 match_value = graceful_chain_get(build_config, *filter_key.split('.'))
                 if filter_value == match_value:
-                    return build_config
+                    build_configs.append(build_config)
+            if build_configs:
```
[the remark this answers is not in the document]
Comment (mlangsdorf) [reply osbs/core.py]: Then items will stay `items = self.get_all_build_configs_by_labels(label_selectors)` and the len(items) > 1 check will fail.

Explicit test case to handle that path added.

[the remark this answers is not in the document]
Comment (twaugh) [reply osbs/core.py]: So if there's only one build config, but it doesn't match the filter criteria, we intend to return it? If so I think the docstring should say that explicitly.

[the remark this answers is not in the document]
Comment (mlangsdorf) [reply osbs/core.py]: No, that's a bug too.

[the remark this answers is not in the document]
Comment (twaugh) [reply osbs/core.py]: Why not just unconditionally `items = build_configs` at the end of the filter block?

If the '!= 1' check is unrelated to whether or not we filtered the results, we should separate them in the code.

[the remark this answers is not in the document]
Comment (mlangsdorf) [reply osbs/core.py]: I had some reason why I didn't want to do it unconditionally, but the code has changed so much that reason is irrelevant.

File: osbs/core.py [hunk 3]
```diff
@@ -357,22 +357,31 @@ def get_build_config_by_labels_filtered(self, label_selectors, filter_key, filte
         Returns a build config matching the given label selectors, filtering against
         another predetermined value. This method will raise OsbsException
         if not exactly one build config is found.
+        if a single build config is found matching the label selectors and no filter value is
+        provided, this method will return that build config.
+        if a single build config is found matching the label selectors but it does doesn't
+        match the filter value, this method will raise OsbsException instead of returning
+        that build config.
```
Comment (twaugh) [inline osbs/core.py]: Could these 5 lines be replaced by adding "after filtering" to the previous sentence, so it ends "if not exactly one build config is found after filtering"?

Comment (mlangsdorf) [reply osbs/core.py]: Fixed.

**Comment by lcarva** [comment]
Updated release notes draft:
> - `get_build_config_by_labels_filtered` raises error if BuildConfig is not found

**Pull request merged**